CRM/PivotReport/DataActivity.php

```diff
@@ -0,0 +1,374 @@
+<?php
+
+/**
+ * Provides a functionality to prepare entity data for Pivot Table.
+ */
+class CRM_PivotReport_DataActivity extends CRM_PivotReport_AbstractData {
+
+  public function __construct($name = NULL) {
+    $this->name = 'Activity';
+  }
+
+  /**
+   * Rebuilds pivot report cache including header and data.
+   *
+   * @param \CRM_PivotCache_AbstractGroup $cacheGroup
+   * @param array $params
+   *
+   * @return array
+   */
+  public function rebuildCache($cacheGroup, array $params) {
+    $this->fields = $this->getFields();
+    $this->emptyRow = $this->getEmptyRow();
+    $this->multiValues = array();
+
+    $time = microtime(true);
+
+    $cacheGroup->clear();
+
+    $count = $this->rebuildData($cacheGroup, $this->name, $params);
+
+    $this->rebuildHeader($cacheGroup, array_merge($this->emptyRow, array(
+      'Activity Date' => null,
+      'Activity Start Date Months' => null,
+      'Activity Expire Date' => null,
+    )));
+
+    return array(
+      array(
+        'rows' => $count,
+        'time' => (microtime(true) - $time),
+      )
+    );
+  }
+
+  /**
+   * Returns an array containing API parameters for Activity 'get' call.
+   *
+   * @param array $inputParams
+   *
+   * @return array
+   */
+  protected function getEntityApiParams(array $inputParams) {
+    $params = array(
+      'sequential' => 1,
+      'is_current_revision' => 1,
+      'is_deleted' => 0,
+      'is_test' => 0,
+      'return' => implode(',', array_keys($this->fields)),
+      'options' => array(
+        'sort' => 'activity_date_time ASC',
+        'limit' => self::ROWS_API_LIMIT,
+      ),
+    );
+
+    $startDate = !empty($inputParams['start_date']) ? $inputParams['start_date'] : NULL;
+    $endDate = !empty($inputParams['end_date']) ? $inputParams['end_date'] : NULL;
+
+    $activityDateFilter = $this->getAPIDateFilter($startDate, $endDate);
+    if (!empty($activityDateFilter)) {
+      $params['activity_date_time'] = $activityDateFilter;
+    }
+
+    return $params;
+  }
+
+  /**
+   * Returns an array containing API date filter conditions basing on specified
+   * dates.
+   *
+   * @param string $startDate
+   * @param string $endDate
+   *
+   * @return array|NULL
+   */
+  private function getAPIDateFilter($startDate, $endDate) {
+    $apiFilter = null;
+
+    if (!empty($startDate) && !empty($endDate)) {
+      $apiFilter = array('BETWEEN' => array($startDate, $endDate));
+    }
+    else if (!empty($startDate) && empty($endDate)) {
+      $apiFilter = array('>=' => $startDate);
+    }
+    else if (empty($startDate) && !empty($endDate)) {
+      $apiFilter = array('<=' => $endDate);
+    }
+
+    return $apiFilter;
+  }
+
+  /**
+   * Returns an array containing $data rows and each row containing multiple values
+   * of at least one field is populated into separate row for each field's
+   * multiple value.
```

Comment:
This description is a bit confusing and I'm not sure I understand what it means, especially the "each row containing multiple values of at least one field is populated into separate row" part. I think that the word "is" should probably be something else